docs: describe how to manage embedded plugin registry for VS Code editor (#2503)
codenvy/che-docs: modules/administration-guide/partials/proc_importing-extensions-that-are-not-in-the-open-vsx-registry.adoc

[id="importing-extensions-that-are-not-in-the-open-vsx-registry"]
= Importing extensions that are not in the Open VSX Registry

This section describes how to add VS Code extensions that are not in the link:https://open-vsx.org/[Open VSX Registry] to {prod-short}'s embedded Open VSX Registry.

> It's possible that link:https://open-vsx.org/[Open VSX Registry] doesn't contain an extension that needs to be used in the workspace. If the extension was not yet published on the link:https://open-vsx.org/[Open VSX Registry]. In that case, please reach out to the extension's author and ask them to publish the extension on the Open VSX registry, the link:https://open-vsx.org/[how to Publish an Extension] docs provide instructions or try to import it into the local registry by using the procedure from this page.

Reviewer: ```suggestion
To use an extension that has not been published to the link:https://open-vsx.org/[Open VSX Registry], contact the extension's author to ask them to publish the extension. See link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[How to Publish an Extension].

If you need the extension and the owner cannot or will not publish it, you may be able to fork it and publish it yourself. Consult the license terms of the extension to find out if this is possible. 

You can also import the extension into the local registry from the procedure in this document. 
```

.Prerequisites

> . The vsix file of the VS Code extension should be deployed somewhere to download it by the direct link.

Reviewer: ```suggestion
. The VS Code extension file (.vsix) must be available to download via direct link.
```

> . The extenshion's id and version that is going to be added.  

Reviewer: ```suggestion
. The ID and version of the extension to be added.  
```

.Procedure

> . Modify `openvsx-sync.json` file in the link:https://github.com/redhat-developer/devspaces/blob/devspaces-3-rhel-8/dependencies/che-plugin-registry/openvsx-sync.json[plugin-registry repository] to add an information about the extension:

Reviewer: ```suggestion
. Modify the plug-in registry's link:https://github.com/redhat-developer/devspaces/blob/devspaces-3-rhel-8/dependencies/che-plugin-registry/openvsx-sync.json[`openvsx-sync.json`] file to add an extension:
```

Reviewer: Note that this is a link to the downstream file... we probably need a new antora variable for this URL so we can pull either https://github.com/eclipse-che/che-plugin-registry/blob/main/openvsx-sync.json or https://github.com/redhat-developer/devspaces/blob/devspaces-3-rhel-8/dependencies/che-plugin-registry/openvsx-sync.json. 

Or should we link to the raw file instead?

https://raw.githubusercontent.com/eclipse-che/che-plugin-registry/main/openvsx-sync.json
vs.
https://raw.githubusercontent.com/redhat-developer/devspaces/devspaces-3-rhel-8/dependencies/che-plugin-registry/openvsx-sync.json

+
[source,json]
----
    {
        "id": "extension_id",
        "download": "url_to_download_vsix_file",
        "version": "extension_version"
    }
----
for example:
+
[source,json]
----
    {
        "id": "redhat.vscode-yaml",
        "version": "1.10.1",
        "download": "https://github.com/redhat-developer/vscode-yaml/releases/download/1.10.1/yaml-1.10.1-19523.vsix"
    }
----

> NOTE: If the extension is available in link:https://marketplace.visualstudio.com/vscode[VS Code Marketplace], the download link could be build like https://extension_publisher.gallery.vsassets.io/_apis/public/gallery/publisher/extension_publisher/extension/extension_name/extension_version/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage. For GoogleCloudTools.cloudcode extension it is https://GoogleCloudTools.gallery.vsassets.io/_apis/public/gallery/publisher/GoogleCloudTools/extension/cloudcode/1.20.3/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage

Author: I'm not sure if it's good to have such info in our docs. 
I've seen this article https://www.eclipse.org/community/eclipse_newsletter/2020/march/1.php
and
```
Downloading extensions from the Microsoft Marketplace for any use other than in Microsoft products is prohibited as well.
```
@nickboldt @l0rd  WDYT?

Reviewer: unless we support / use this method, then +1, we shouldn't mention it.

And if it violates M$'s license terms, then +100, we should definitely remove it.

Reviewer 2: I think that we can mention how to retrieve the download URL of an extension published in the microsoft marketplace (there is a download link in the msft marketplace anyway). We should warn that downloading the extension is prohibited but I don't see anything that confirms that in the [Microsoft Visual Studio Marketplace Terms of Use](https://aka.ms/vsmarketplace-ToU) (maybe that has changed in the meantime?).

Reviewer: I definitely recall a "thou shalt only use these extensions with our products" caveat being on the site at some point, but http://aka.ms/VSMarketplace-TOU (last updated June 2021) doesn't have that anywhere. Also nothing in an older version https://cdn.vsassets.io/v/M146_20190123.39/_content/Microsoft-Visual-Studio-Marketplace-Terms-of-Use.pdf from Jan 2019. 

Maybe it was somewhere else, like on https://github.com/VSCodium/vscodium/blob/master/DOCS.md#extensions--marketplace which says "Being a vscode based editor, VSCodium gets additional features by installing VS Code extensions. Unfortunately, as Microsoft prohibits usages of the Microsoft marketplace by any other products or redistribution of .vsix files from it, in order to use VS Code extensions in non-Microsoft products those need to be installed differently." and links to https://github.com/microsoft/vscode/issues/31168#issuecomment-606757206 and https://github.com/microsoft/vscode/issues/31168#issuecomment-1136375670, which states in summary:

```
Since Microsoft won't directly interpret their Visual Studio Marketplace Terms of Use for you, I will since they're quite clear:

Short Answer: NO, YOU CANNOT LEGALLY USE THE MICROSOFT MARKETPLACE IN ANY NON-MICROSOFT PRODUCT.
```

See also https://github.com/microsoft/vscode/wiki/Differences-between-the-repository-and-Visual-Studio-Code

Reviewer: Given the [whois data](https://who.is/whois/vsassets.io) for vsassets.io shows it's owned by M$, I would assume that pulling files from there is the same as pulling files from https://marketplace.visualstudio.com/ and therefore would violate their TOU. 

-1 to link to asset files. If people want to figure it out themselves and violate the TOU, that's up to them. We should not document how to make it easier.

Reviewer 2: Ok so let's not mention the direct URL to the Microsoft marketplace @svor

Reviewer: fixed in PR

> NOTE: If the extension is available in link:https://marketplace.visualstudio.com/vscode[VS Code Marketplace], the download link could be build like https://extension_publisher.gallery.vsassets.io/_apis/public/gallery/publisher/extension_publisher/extension/extension_name/extension_version/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage. For GoogleCloudTools.cloudcode extension it is https://GoogleCloudTools.gallery.vsassets.io/_apis/public/gallery/publisher/GoogleCloudTools/extension/cloudcode/1.20.3/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage

Reviewer: ```suggestion

```

> . Build the container image with plugin registry and push the image to a container registry:

Reviewer: ```suggestion
. To build the plug-in registry container image and publish it to a container registry, run the following commands:
```

+
[subs="+attributes,+quotes"]
----

> $ ./build.sh -o username -r quay.io -t next

Reviewer: ```suggestion
$ ./build.sh -o username -r quay.io -t custom
```

> $ {docker-cli} push username/che-theia-devspaces:next

Reviewer: ```suggestion
$ {docker-cli} push quay.io/username/che-plugin-registry:custom
```

----
. Customize an instance of `CheCluster` Custom Resource in the cluster to point the image and save the changes:
+
[source,yaml,subs="+quotes"]
----
spec:
  components:
    pluginRegistry:
      deployment:
        containers:

>           - image: quay.io/username/pluginregistry-rhel8:next

Reviewer: ```suggestion
          - image: quay.io/username/plugin-registry:custom
```

      openVSXURL: '' 
----

> . Check that `che-plugin-registry` pod restarted and it's running. Restart the workspace and check plugins in the Plugins view.

Reviewer: ```suggestion
. Check that `plugin-registry` pod restarted and is running. Restart the workspace and check the available extensions in the Extensions view.
```

[id="importing-extensions-that-are-not-in-the-open-vsx-registry"]
= Importing extensions that are not in the Open VSX Registry

> This section describes how to add VS Code extensions that are not in the link:https://open-vsx.org/[Open VSX Registry] to {prod-short}'s embedded Open VSX Registry.

Reviewer: ```suggestion
This section describes how to add Visual Studio Code extensions that are not in the link:https://open-vsx.org/[Open VSX Registry] to {prod-short}'s embedded Open VSX Registry.
```

To use an extension that has not been published to the link:https://open-vsx.org/[Open VSX Registry], contact the extension's author to ask them to publish the extension. See link:https://github.com/eclipse/openvsx/wiki/Publishing-Extensions#how-to-publish-an-extension[How to Publish an Extension].

> If you need the extension and the owner cannot or will not publish it, you may be able to fork it and publish it yourself. Consult the license terms of the extension to find out if this is possible. 

Reviewer: ```suggestion
If you need the extension and the owner cannot or will not publish it, you might be able to fork it and publish it yourself. Consult the license terms of the extension to find out if this is possible. 
```

You can also import the extension into the local registry from the procedure in this document. 

.Prerequisites

> . The VS Code extension file (.vsix) must be available to download via direct link.

Reviewer: ```suggestion
. The Visual Studio Code extension file (.vsix) must be available to download from a valid URL.
```

. The ID and version of the extension to be added.  

.Procedure
. Modify `openvsx-sync.json` file in the link:https://github.com/redhat-developer/devspaces/blob/devspaces-3-rhel-8/dependencies/che-plugin-registry/openvsx-sync.json[plugin-registry repository] to add an information about the extension:
+
[source,json]
----
    {
        "id": "extension_id",
        "download": "url_to_download_vsix_file",
        "version": "extension_version"
    }
----
for example:
+
[source,json]
----
    {
        "id": "redhat.vscode-yaml",
        "version": "1.10.1",
        "download": "https://github.com/redhat-developer/vscode-yaml/releases/download/1.10.1/yaml-1.10.1-19523.vsix"
    }
----

> . To build the plug-in registry container image and publish it to a container registry, run the following commands:

Reviewer: ```suggestion
. To build the plugin registry container image and publish it to a container registry, run the following commands:
```

If you need the extension and the owner cannot or will not publish it, you may be able to fork it and publish it yourself. Consult the license terms of the extension to find out if this is possible. 

You can also import the extension into the local registry from the procedure in this document. 

.Prerequisites
. The VS Code extension file (.vsix) must be available to download via direct link.
. The ID and version of the extension to be added.  

.Procedure
. Modify `openvsx-sync.json` file in the link:https://github.com/redhat-developer/devspaces/blob/devspaces-3-rhel-8/dependencies/che-plugin-registry/openvsx-sync.json[plugin-registry repository] to add an information about the extension:
+
[source,json]
----
    {
        "id": "extension_id",
        "download": "url_to_download_vsix_file",
        "version": "extension_version"
    }
----
for example:
+
[source,json]
----
    {
        "id": "redhat.vscode-yaml",
        "version": "1.10.1",
        "download": "https://github.com/redhat-developer/vscode-yaml/releases/download/1.10.1/yaml-1.10.1-19523.vsix"
    }
----

> NOTE: If the extension is available in link:https://marketplace.visualstudio.com/vscode[VS Code Marketplace], the download link could be build like https://extension_publisher.gallery.vsassets.io/_apis/public/gallery/publisher/extension_publisher/extension/extension_name/extension_version/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage. For GoogleCloudTools.cloudcode extension it is https://GoogleCloudTools.gallery.vsassets.io/_apis/public/gallery/publisher/GoogleCloudTools/extension/cloudcode/1.20.3/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage

Reviewer: ```suggestion

```
